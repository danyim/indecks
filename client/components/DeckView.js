import React from 'react';
import Card from './Card';
import DeckNavigator from './DeckNavigator';

const DeckView = React.createClass({
  render() {
    const { deckId, cardIndex } = this.props.params;
    // index of the deck
    const i = this.props.decks.findIndex((deck) => deck.id === deckId)
    // get us the post
    const deck = this.props.decks[i];

    return (
      <div>
        <div className="single-card">
          <Card card={deck.cards[cardIndex - 1]} />
        </div>
        <DeckNavigator deck={deck} cardIndex={parseInt(cardIndex)} mode="view" />
      </div>
    )
  }
});

export default DeckView;
