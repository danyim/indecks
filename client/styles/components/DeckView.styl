@import '../_variables';

.deck-view
  width 100%
  max-width pageWidth
  // .grid-figure
  //   flex-basis calc(50% - 2rem)
  //   margin 0 1rem 2rem 0rem
.title-card
  border 1px solid lightgrey
  background-color white
  border-bottom 5px solid black
  margin 0 1rem 2rem 1rem
  padding 2rem
  padding-top 4rem
  flex-direction row
  flex-wrap wrap
.title-text
  margin-top 2rem
  flex 1 1 75%
  text-overflow ellipsis
  *
    margin-right 2rem
    flex 1 1 auto
.control-buttons
  width 25%
  flex-direction column
  flex 1 0 25%
  button
    flex 0 0 auto
