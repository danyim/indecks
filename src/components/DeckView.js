import React from 'react';
import Card from './Card';
import DeckNavigator from './DeckNavigator';
import { Link, browserHistory } from 'react-router';
import ExportDeckButton from './ExportDeckButton'
import slug from 'slug';
import styles from '../styles/components/DeckView';

const DeckView = React.createClass({
  // Parses all the cards of a deck object and performs a special replacement
  mdReplacer(k, v) {
    if(k === 'cards') {
      const cards = [...v];
      for(let card of cards) {
        // card.description = 'This is a test';
      }
      return cards;
    }
    return v;
  },

  handleEditCard(deckId, i) {
    browserHistory.push(`/edit/${deckId}/${i + 1}`);
  },

  handleRemoveDeck(deckId) {
    if(confirm('Are you sure you want to delete this deck?')) {
      this.props.removeDeck(deckId);
      browserHistory.push(`/`);
    }
  },

  render() {
    const { deckId } = this.props.params;
    // index of the deck
    const i = this.props.decks.findIndex((deck) => deck.id === deckId)
    // get us the post
    const deck = this.props.decks[i];
    const mode = 'view';

    const emptyMsg = (() => {
      if(deck.cards.length === 0 ) {
        return <p>Click the + button on the top left to add a card</p>
      }
    })()

    return (
      <section className={`${styles['deck-view']}`}>
        <div className={`${styles['title-card']}`}>
          <div className={`${styles['title-text']}`}>
            <h1>{deck.title}</h1>
            <p>{deck.description}</p>
          </div>
          <div className={`${styles['control-buttons']}`}>
            <Link className="button" to={`/view/${deck.id}/1`}>
              <span className={`${styles['comment-count']}`}>
                Play Deck
              </span>
            </Link>
            <Link className="button" to={`/view/${deck.id}/1`}>
              <span className={`${styles['comment-count']}`}>
                Edit Details
              </span>
            </Link>
            {/*<Link className="button" to={`/edit/${deck.id}/1`}>
              <span className={`${styles['comment-count']}`}>
                Edit
              </span>
            </Link>*/}
            <ExportDeckButton
              filename={`${slug(deck.title)}.json`}
              label="Export Deck (JSON)"
              className="button"
              style={{}}
              /*
                // Use this if you want to use the replacer
                exportFile={() =>  JSON.stringify(deck, this.mdReplacer, 2)}
              */
             exportFile={() => JSON.stringify(deck, null, 2)} />
            <a className="button btn-delete" onClick={() => this.handleRemoveDeck(deck.id)}>
                Delete Deck
            </a>
          </div>
        </div>

        <div className={`wrap-row ${styles['grid']}`}>
          {
            deck.cards.map((c, i) =>
              <Card card={c} key={i} handleOnClick={() => this.handleEditCard(deckId, i)} />
            )
          }
          {emptyMsg}
        </div>
      </section>
    )
  }
});

export default DeckView;
