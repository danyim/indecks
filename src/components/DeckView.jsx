import React from 'react';
import Card from './Card';
import DeckNavigator from './DeckNavigator';
import { Link, browserHistory } from 'react-router';
import ExportDeckButton from './ExportDeckButton';
import slug from 'slug';
import styles from '../styles/components/DeckView';

const propTypes = {};

const defaultProps = {};

class DeckView extends React.Component {
  constructor(props) {
    super(props);
  }

  // Parses all the cards of a deck object and performs a special replacement
  mdReplacer(k, v) {
    if(k === 'cards') {
      const cards = [...v];
      for(let card of cards) {
        // card.description = 'This is a test';
      }
      return cards;
    }
    return v;
  }

  handleOnCardClick(deckId, i) {
    browserHistory.push(`/edit/${deckId}/${i + 1}`);
  }

  render() {
    const emptyMsg = (() => {
      if(this.props.deck.cards.length === 0 ) {
        return <p className="center">Click the + button on the top left to add a card</p>
      }
    })()

    return (
      <section className={`${styles['deck-view']}`}>
        <div className={`${styles['title-card']}`}>
          <div className={`${styles['title-text']}`}>
            <h1>{this.props.deck.title}</h1>
            <p>{this.props.deck.description}</p>
            <p><strong>{`${this.props.deck.cards.length} cards in this deck`}</strong></p>
          </div>
          <div className={`${styles['control-buttons']}`}>
            <Link className="button" to={`/view/${this.props.deck.id}/1`}>
              <span>
                Play Deck
              </span>
            </Link>
            <Link className="button" to={`/edit/${this.props.deck.id}`}>
              <span>
                Edit Details
              </span>
            </Link>
            <ExportDeckButton
              filename={`${slug(this.props.deck.title)}.json`}
              label="Export Deck"
              className="button"
              style={{}}
              /*
                // Use this if you want to use the replacer
                exportFile={() =>  JSON.stringify(deck, this.mdReplacer, 2)}
              */
             exportFile={() => JSON.stringify(this.props.deck, null, 2)} />
            <a className="button btn-delete" onClick={() => this.props.handleRemoveDeck(this.props.deck.id)}>
                Delete Deck
            </a>
          </div>
        </div>

        <div className={`wrap-row ${styles['grid']}`}>
          {
            this.props.deck.cards.map((c, i) =>
              <Card card={c} key={i} handleOnClick={() => this.handleOnCardClick(this.props.deck.id, i)} />
            )
          }
          {emptyMsg}
        </div>
      </section>
    )
  }
}

DeckView.propTypes = propTypes;
DeckView.defaultProps = defaultProps;

export default DeckView;
