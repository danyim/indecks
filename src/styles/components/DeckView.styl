@import '../_variables';

.deck-view
  width 100%
  max-width pageWidth
  // .grid-figure
  //   flex-basis calc(50% - 2rem)
  //   margin 0 1rem 2rem 0rem

// Individual card styles
.grid
  justify-content center
  figure
    flex-basis calc(50% - 4rem)
  // div > h1
  //   width 550px
  //   // white-space nowrap
  //   overflow hidden
  //   text-overflow ellipsis

@media (max-width: cardWidth)
  .grid figure
    flex-basis calc(100% - 4rem)

.title-card
  border 1px solid lightgrey
  background-color white
  border-bottom 5px solid black
  margin 0 1rem 2rem 1rem
  padding 2rem
  padding-top 4rem
  flex-flow row wrap
.title-text
  margin-top 2rem
  flex 1 1 75%
  order 1
  text-overflow ellipsis
  *
    margin-right 2rem
    flex 1 1 auto
.control-buttons
  flex-flow column nowrap
  flex-grow 1
  order 2
  transition all 0.2s ease-in
  padding-left 1.0rem
  button, a
    flex 1 0 calc(25% - 1rem)
    margin-top 0.6rem
    max-height 4.0rem
  button:first-child, a:first-child
    margin-top 0

@media (max-width: cardWidth)
  .title-card
    flex-flow column nowrap
    justify-content: center
  .control-buttons
    flex-flow row nowrap
    padding-left 0
    button, a
      flex 1 0 calc(25% - 2rem)
      margin-top 0
      margin-left 0.6rem
    button:first-child, a:first-child
      margin-left 0

