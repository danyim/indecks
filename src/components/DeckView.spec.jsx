import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { browserHistory } from 'react-router'
import DeckView from './DeckView'

const defaultProps = {
  deck: {
    id: 'ABCDEFG',
    title: 'Deck title',
    description: 'Deck description',
    cards: [
      {
        index: 1,
        title: 'Card',
        answer: 'Card description'
      }
    ]
  },
  maxDeckTitleLength: 160,
  maxDeckDescLength: 300,
  handleDuplicateCard: jest.fn(),
  handleEditDeck: jest.fn(),
  handleRemoveCard: jest.fn(),
  handleRemoveDeck: jest.fn()
}

function setup (props = defaultProps) {
  const wrapper = shallow(<DeckView {...props} />)

  return {
    props,
    wrapper
  }
}

function setupFull (props = defaultProps) {
  const wrapper = mount(<DeckView {...props} />)

  return {
    props,
    wrapper
  }
}

describe('DeckView', () => {
  it('should render self and subcomponents', () => {
    // const { wrapper } = setup();

    const tree = renderer.create(
      <DeckView {...defaultProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render subcomponents when there are cards in the deck', () => {
    const { props, wrapper } = setup()

    expect(wrapper.find('Card').length).toEqual(props.deck.cards.length)
    expect(wrapper.find('Overlay').length).toEqual(props.deck.cards.length)
  })

  it('should render a notice if no cards are found', () => {
    const { wrapper } = setup({
      ...defaultProps,
      deck: {
        ...defaultProps.deck,
        cards: []
      }
    })
    expect(wrapper.find('p.center').text()).toEqual('Click the + button on the top left to add a card')
  })

  it('should navigate to the VIEW card route when the "View" card overlay button is clicked', () => {
    spyOn(browserHistory, 'push')
    const { props, wrapper } = setupFull()

    wrapper.find('Overlay button[children="View"]').first().simulate('click')
    expect(browserHistory.push).toHaveBeenCalledWith(`/view/${props.deck.id}/1`)
  })

  it('should navigate to the EDIT card route when the "Edit" card overlay button is clicked', () => {
    spyOn(browserHistory, 'push')
    const { props, wrapper } = setupFull()

    wrapper.find('Overlay button[children="Edit"]').first().simulate('click')
    expect(browserHistory.push).toHaveBeenCalledWith(`/edit/${props.deck.id}/1`)
  })

  it('should call the handleRemoveCard prop when the "Delete" card overlay button is clicked', () => {
    const handler = jest.fn()
    spyOn(window, 'confirm').and.returnValue(true)
    spyOn(browserHistory, 'push')
    const { props, wrapper } = setup({
      ...defaultProps,
      handleRemoveCard: handler
    })

    wrapper.find('Overlay button[children="Delete"]').first().simulate('click')
    expect(handler.mock.calls.length).toBe(1)
    expect(browserHistory.push).toHaveBeenCalledWith(`/view/${props.deck.id}`)
  })

  it('should call the handleDuplicateCard prop when the "Duplicate" card overlay button is clicked', () => {
    const handler = jest.fn()
    const { wrapper } = setup({
      ...defaultProps,
      handleDuplicateCard: handler
    })

    wrapper.find('button[children="Duplicate"]').first().simulate('click')
    wrapper.find('button[children="Duplicate"]').last().simulate('click')
    expect(handler.mock.calls.length).toBe(2)
  })

  it('should call the handleRemoveDeck prop when the "Delete Deck" button is clicked', () => {
    const handler = jest.fn()
    const { wrapper } = setup({
      ...defaultProps,
      handleRemoveDeck: handler
    })

    wrapper.find('button[children="Delete Deck"]').first().simulate('click')
    expect(handler.mock.calls.length).toBe(1)
  })

  it('should call the handleEditDeck prop when the title changes', () => {
    const handler = jest.fn()
    const { wrapper } = setupFull({
      ...defaultProps,
      handleEditDeck: handler
    })

    const rieInput = wrapper.find('RIEInput')
    rieInput.simulate('click') // Go into editing mode
    const input = wrapper.find('input.large') // Find the <input> DOM
    input.simulate('input', { target: { value: 'Test' } }) // Trigger the onInput
    input.node.value = 'Test' // Explicitly set the value to the test value
    input.simulate('blur') // Get out of editing mode
    expect(handler.mock.calls.length).toBe(1)
  })

  it('should call the handleEditDeck prop when the description changes', () => {
    const handler = jest.fn()
    const { wrapper } = setupFull({
      ...defaultProps,
      handleEditDeck: handler
    })

    const rieInput = wrapper.find('RIETextArea')
    rieInput.simulate('click') // Go into editing mode
    const input = wrapper.find('textarea.paragraph') // Find the <input> DOM
    input.simulate('input', { target: { value: 'Test' } }) // Trigger the onInput
    input.node.value = 'Test' // Explicitly set the value to the test value
    input.simulate('blur') // Get out of editing mode
    expect(handler.mock.calls.length).toBe(1)
  })

  it('should show an invalidation when a 0-char deck title is entered', () => {
    const { wrapper } = setupFull()
    const rieInput = wrapper.find('RIEInput').first()

    rieInput.simulate('click')
    const input = wrapper.find('input.large')
    input.simulate('input', {target: {value: ''}})
    expect(input.hasClass('invalid')).toEqual(true)
  })

  it('should show an invalidation when large deck title is entered', () => {
    const { wrapper } = setupFull({
      ...defaultProps,
      maxDeckTitleLength: 4
    })
    const rieInput = wrapper.find('RIEInput')
    rieInput.simulate('click')
    const input = wrapper.find('input.large')
    input.simulate('input', {target: {value: '12345'}})
    expect(input.hasClass('invalid')).toEqual(true)
  })

  it('should show an invalidation when a 0-char deck description is entered', () => {
    const { wrapper } = setupFull()
    const rieTextArea = wrapper.find('RIETextArea')

    rieTextArea.simulate('click')
    const input = wrapper.find('textarea.paragraph textarea.editable')
    input.simulate('input', {target: {value: ''}})
    expect(input.hasClass('invalid')).toEqual(true)
  })

  it('should show an invalidation when a large deck description is entered', () => {
    const { wrapper } = setupFull({
      ...defaultProps,
      maxDeckDescLength: 4
    })
    const rieTextArea = wrapper.find('RIETextArea')

    rieTextArea.simulate('click')
    const input = wrapper.find('textarea.paragraph textarea.editable')
    input.simulate('input', {target: {value: '12345'}})
    expect(input.hasClass('invalid')).toEqual(true)
  })

  it('should convert the deck into JSON when the export button is clicked', () => {
    const { props, wrapper } = setupFull()

    const exportButton = wrapper.find('ExportDeckButton')
    expect(exportButton.prop('exportFile')()).toEqual(JSON.stringify(props.deck, null, 2))
  })
})
