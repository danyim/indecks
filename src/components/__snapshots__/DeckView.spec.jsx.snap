// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckView should render self and subcomponents 1`] = `
<section
  className="deck-view"
>
  <div
    className="title-card"
  >
    <div
      className="title-text"
    >
      <RIEInput
        change={[Function]}
        classInvalid="invalid"
        classLoading="loading"
        className="large editable"
        maxLength={160}
        minLength="1"
        placeholder="Click here to add a title"
        propName="title"
        validate={[Function]}
        value="Deck title"
      />
      <p>
        <strong>
          1 cards
        </strong>
         
        <ModalHelpButton
          style={
            Object {
              "minWidth": "350px",
            }
          }
        >
          <p>
            Click anywhere on the deck's title or description to edit the field. Your changes are saved immediately.
          </p>
        </ModalHelpButton>
      </p>
      <RIETextArea
        change={[Function]}
        classInvalid="invalid"
        classLoading="loading"
        className="paragraph editable m-t"
        maxLength={300}
        minLength="1"
        placeholder="Click here to add a description"
        propName="description"
        rows="6"
        validate={[Function]}
        value="Deck description"
      />
       
    </div>
    <div
      className="control-buttons"
    >
      <Link
        className="button outline"
        replace={false}
        to="/view/ABCDEFG/1"
      >
        Play Deck  
        <i
          className="fa fa-play-circle"
        />
      </Link>
      <ExportDeckButton
        className="button"
        disabled={false}
        exportFile={[Function]}
        filename="Deck-title.json"
        label="Export Deck"
        style={Object {}}
      />
      <button
        className="btn-delete"
        onClick={[Function]}
      >
        Delete Deck
      </button>
    </div>
  </div>
  <div
    className="wrap-row grid"
  >
    <Card
      card={
        Object {
          "answer": "Card description",
          "index": 1,
          "title": "Card",
        }
      }
      className="card-contents"
      flipped={false}
      handleOnClick={[Function]}
      trimOverflow={true}
      trimOverflowLength={125}
    >
      <Overlay
        className=""
      >
        <OverlayRow
          className=""
        >
          <button
            className="button"
            onClick={[Function]}
          >
            View
          </button>
          <button
            className="button"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button btn-delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </OverlayRow>
        <OverlayRow
          className=""
        >
          <button
            className="button"
            onClick={[Function]}
          >
            Duplicate
          </button>
        </OverlayRow>
      </Overlay>
    </Card>
  </div>
</section>
`;
